File: Assets/Game/Scripts/Moduls/P_Modul_30.lua
local ObjectStateBackupUtils = Ballance2.Utils.ObjectStateBackupUtils

---P_Modul_30
---跷跷板机关
---@class P_Modul_30 : ModulBase
---@field P_Modul_30_Wippe PhysicsObject
P_Modul_30 = ModulBase:extend()

function P_Modul_30:new()
  P_Modul_30.super.new(self)
  self.EnableBallRangeChecker = true
  self.BallCheckeRange = 60
end

function P_Modul_30:Active()
  ModulBase.Active(self)
  self.P_Modul_30_Wippe.CollisionID = GamePlay.BallSoundManager:GetSoundCollIDByName('Wood')
  self.P_Modul_30_Wippe:Physicalize()
end
function P_Modul_30:Deactive()
  self.P_Modul_30_Wippe:UnPhysicalize(true)
  ModulBase.Deactive(self)
end

function P_Modul_30:ActiveForPreview()
  self.gameObject:SetActive(true)
end
function P_Modul_30:DeactiveForPreview()
  self.gameObject:SetActive(false)
end

function P_Modul_30:Reset()
  ObjectStateBackupUtils.RestoreObject(self.gameObject)
end
function P_Modul_30:Backup()
  ObjectStateBackupUtils.BackUpObject(self.gameObject)
end
function P_Modul_30:BallEnterRange()
  if not self.IsPreviewMode and self.IsActive then
    self.P_Modul_30_Wippe:WakeUp()
  end
end

function CreateClass:P_Modul_30()
  return P_Modul_30()
end

File: Assets/Game/GamePlay/Scripts/Moduls/P_Modul_30.cs
using Ballance2.Utils;
using BallancePhysics.Wapper;

namespace Ballance2.Game.GamePlay.Moduls
{
  public class P_Modul_30 : ModulBase 
  {
    public PhysicsObject P_Modul_30_Wippe;

    public P_Modul_30() {
      EnableBallRangeChecker = true;
      BallCheckeRange = 60;
    }

    public override void Active()
    {
      base.Active();
      P_Modul_30_Wippe.CollisionID = GamePlayManager.Instance.BallSoundManager.GetSoundCollIDByName("Wood");
      P_Modul_30_Wippe.gameObject.SetActive(true);
      P_Modul_30_Wippe.Physicalize();
    }

    public override void Deactive()
    {
      P_Modul_30_Wippe.UnPhysicalize(true);
      base.Deactive();
    }

    public override void ActiveForPreview()
    {
      gameObject.SetActive(true);
    }

    public override void DeactiveForPreview()
    {
      gameObject.SetActive(false);
    }

    public override void Reset(ModulBaseResetType type)
    {
      ObjectStateBackupUtils.RestoreObjectAndChilds(gameObject);
    }

    public override void Backup()
    {
      ObjectStateBackupUtils.BackUpObjectAndChilds(gameObject);
    }

    protected override void BallEnterRange()
    {
      if (!IsPreviewMode && IsActive)
        P_Modul_30_Wippe.WakeUp();
    }
  }
}